R/sensitivityCalculation.R

```diff
@@ -0,0 +1,118 @@
+#' @name sensitivityCalculation
+#' @title Carry out and visualize sensitivity analysis (with OSPSuite)
+#'
+#' @param simulation An object of type `Simulation`.
+#' @param outputPaths Path (or a vector of paths) to the output(s) for which the
+#'   sensitivity will be analyzed.
+#' @param parameterPaths A single or a vector of the parameter path(s) to be
+#'   varied.
+#' @param variationRange Optional numeric vector defining the scaling of the
+#'   parameters. The same variation range should be applied to all specified
+#'   parameters. If not specified, the following vector will be used: c(0.1,
+#'   0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 2, 3, 4, 5,  6, 7, 8, 9, 10).
+#' @param pkParameters A vector of names of PK parameters for which the
+#'   sensitivities should be calculated. For a full set of available PK
+#'   parameters, run `names(ospsuite::StandardPKParameter)`. By default, only
+#'   the following parameters will be considered: `"C_max"`, `"t_max"`,
+#'   `"AUC_inf"`. If `NULL`, all available PK-parameters will be calculated.
+#' @param pkDataFilePath Path to spreadsheets in which
+#'   PK-parameter data should be saved. Default is `NULL`,
+#'   meaning the data will not be saved to a spreadsheet.
+#'
+#' @note
+#'
+#' - PK parameter `"t_max"` in some contexts may not show any variation.
+#'
+#' @examples
+#'
+#' library(ospsuite)
+#'
+#' simPath <- system.file("extdata", "Aciclovir.pkml", package = "esqlabsR")
+#' simulation <- loadSimulation(simPath)
+#' outputPaths <- "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
+#' parameterPaths <- c(
+#'   "Aciclovir|Lipophilicity",
+#'   "Applications|IV 250mg 10min|Application_1|ProtocolSchemaItem|Dose",
+#'   "Neighborhoods|Kidney_pls_Kidney_ur|Aciclovir|Glomerular Filtration-GFR|GFR fraction"
+#' )
+#'
+#' # extract the results into a list of dataframes
+#' sensitivityCalculation(
+#'   simulation = simulation,
+#'   outputPaths = outputPaths,
+#'   parameterPaths = parameterPaths
+#' )
+#' @export
+
+sensitivityCalculation <- function(simulation,
+                                   outputPaths,
+                                   parameterPaths,
+                                   variationRange = c(seq(0.1, 1, by = 0.1), seq(2, 10, by = 1)),
+                                   pkParameters = c("C_max", "t_max", "AUC_inf"),
+                                   pkDataFilePath = NULL) {
+  # fail fast if specified PK parameters are not part of standard `enum`
+  validateIsIncluded(pkParameters, names(ospsuite::StandardPKParameter))
+
+  # set outputs to the provided path
+  clearOutputs(simulation)
+  addOutputs(outputPaths, simulation)
+
+  # save raw names of paths to create named list
+  parameterPathNames <- parameterPaths
+
+  # create simulation batch for efficient calculations
+  parameterPaths <- getAllParametersMatching(parameterPaths, simulation)
+
+  # extract dataframes with results
+  batchResults <- purrr::map(
```

Comment:
It is not better _per se_. I am just trying to be internally consistent because I use `map2` in other place, and so I want to consistently use `map` family everywhere I am working with vectors.

As an aside, since _ospsuite_ has _purrr_ has a hard dependency, the _esqlabsR_ package also relies on _purrr_, so we are not gaining a new dependency. 

Here is a short summary of why I personally prefer `map` function family over `apply` function family.

Ref. https://r4ds.had.co.nz/iteration.html#base-r

![image](https://user-images.githubusercontent.com/11330453/155160713-6240bab5-7470-4de7-908c-72ee5644eede.png)